carl/context/context_space.py

```diff
@@ -0,0 +1,133 @@
+        context = {cf.name: cf.default_value for cf in self.context_space.values()}
+        return context
+    
+    def to_gymnasium_space(
+        self, context_feature_names: List[str] | None = None, as_dict: bool = False
+    ) -> spaces.Space:
+        if context_feature_names is None:
+            context_feature_names = self.context_feature_names
+        if as_dict:
+            context_space = {}
+
+            for cf_name in context_feature_names:
+                context_feature = self.context_space[cf_name]
+                if isinstance(context_feature, NumericalContextFeature):
+                    context_space[context_feature.name] = spaces.Box(low=context_feature.lower, high=context_feature.upper)
+                else:
+                    raise ValueError(
+                        f"Context features must be of type NumericalContextFeature."
+                        f"Got {type(context_feature)}."
+                    )
+            print(context_space)
+            return spaces.Dict(context_space)
+        else:
+            low = np.array(
+                [self.context_space[cf].lower for cf in context_feature_names]
+            )
+            high = np.array(
+                [self.context_space[cf].upper for cf in context_feature_names]
+            )
+
+            return spaces.Box(low=low, high=high, dtype=np.float32)
+
+
+class ContextSampler(ConfigurationSpace):
+    def __init__(
+        self,
+        context_distributions: list[ContextFeature] | str | DictConfig,
+        context_space: ContextSpace,
+        seed: int,
+        name: str | None = None,
+    ):
+        self.context_distributions = context_distributions
+        super().__init__(name=name, seed=seed)
+
+        if isinstance(context_distributions, list):
+            self.add_context_features(context_distributions)
+        elif type(context_distributions) in [str, DictConfig]:
+            cs = search_space_to_config_space(context_distributions)
+            self.add_context_features(cs.get_hyperparameters())
+        else:
+            raise ValueError(f"Unknown type `{type(context_distributions)}` for `context_distributions`.")
+        
+        self.context_feature_names = [cf.name for cf in context_distributions]
+        self.context_space = context_space
+
+        self.verify_distributions(context_distributions=context_distributions, context_space=context_space)
+
+    def add_context_features(self, context_features: list[ContextFeature]) -> None:
+        self.add_hyperparameters(context_features)
+
+    @staticmethod
+    def verify_distributions(context_distributions: list[ContextFeature], context_space: ContextSpace) -> bool:
+        # TODO verify distributions or context?
+        return True
+
+    def sample_contexts(self, n_contexts: int) -> Contexts:
+        contexts = self._sample_contexts(size=n_contexts)
+
+        # Convert to dict
+        contexts = {i: self.insert_defaults(C) for i, C in enumerate(contexts)}
```

Comment:
Do we need the defaults in there at all? We override the defaults anyway when we pass the context to the env, no? I think having all values in here will look confusing to users

Reply:
In that case it could be optional... though, where do you actually want that? I can't really think of any cases off the top of my head